Set the traffic class of the NW WebRTC backend based on WebRTC backend DSCP information

youennf opened the pull request:
#### 51c4e45400a9953900408217cb4ac5b9a74f5f86
<pre>
Set the traffic class of the NW WebRTC backend based on WebRTC backend DSCP information
<a href="https://bugs.webkit.org/show_bug.cgi?id=254072">https://bugs.webkit.org/show_bug.cgi?id=254072</a>
rdar://106722223

Reviewed by Alex Christensen.

Set traffic class based on provided DSCP values for TCP and UDP nw sockets.

* Source/WebKit/NetworkProcess/webrtc/NetworkRTCTCPSocketCocoa.mm:
(WebKit::NetworkRTCTCPSocketCocoa::setOption):
* Source/WebKit/NetworkProcess/webrtc/NetworkRTCUDPSocketCocoa.mm:
(WebKit::NetworkRTCUDPSocketCocoaConnections::setOption):
* Source/WebKit/NetworkProcess/webrtc/NetworkRTCUtilitiesCocoa.h:
* Source/WebKit/NetworkProcess/webrtc/NetworkRTCUtilitiesCocoa.mm:
(WebKit::trafficClassFromDSCP):

Canonical link: <a href="https://commits.webkit.org/261917@main">https://commits.webkit.org/261917@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/57ecfde590891a329c0ddf0bb29227de61aacd30

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/112891 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/22051 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/1569 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/4672 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/121391 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/23399 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/13216 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/5866 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/118678 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/17381 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/100641 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/105983 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/99341 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/1179 "4 failures") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/46400 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/14348 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/1219 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/15050 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/10551 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/20364 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/53211 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8330 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/16900 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/NetworkProcess/webrtc/NetworkRTCTCPSocketCocoa.mm:
```diff
@@ -139,9 +139,18 @@ static inline void processIncomingData(RetainPtr<nw_connection_t>&& nwConnection
+    if (option != rtc::Socket::OPT_DSCP)
+        return;
+
+    auto trafficClass = trafficClassFromDSCP(static_cast<rtc::DiffServCodePoint>(value));
```
Can the parameter type be an rtc::DiffServCodePoint instead of an int?

youennf replied:
DiffServCodePoint is only if option is OPT_DSCP.
In the future, there might be other option values we want to use.

achristensen07 commented on Source/WebKit/NetworkProcess/webrtc/NetworkRTCUDPSocketCocoa.mm:
```diff
@@ -240,6 +241,8 @@ static inline bool isNat64IPAddress(const rtc::IPAddress& ip)
         auto connectionStateTracker = ConnectionStateTracker::create();
         protectedThis->setupNWConnection(nwConnection, connectionStateTracker.get(), remoteAddress);
+        if (protectedThis->m_trafficClass)
+            nw_connection_reset_traffic_class(nwConnection, *protectedThis->m_trafficClass);
```
Can we use nw_parameters_set_traffic_class also in this case before starting the connection, instead of starting the connection then calling nw_connection_reset_traffic_class immediately upon start?

youennf replied:
This connection is created from receiving packets, we do not have any nw_parameters in this code path.

achristensen07 commented on Source/WebKit/NetworkProcess/webrtc/NetworkRTCUtilitiesCocoa.mm:
```diff
@@ -80,6 +80,42 @@ bool isKnownTracker(const WebCore::RegistrableDomain& domain)
+        return SO_TC_BK_SYS;
+    case rtc::DiffServCodePoint::DSCP_AF41:
+        return SO_TC_VI;
+    case rtc::DiffServCodePoint::DSCP_AF42:
```
This could be combined with the previous one.  Why don't we want to do anything for DSCP_AF43?

youennf replied:
Will do.
The mapping of priority to DSCP is defined in https://datatracker.ietf.org/doc/html/rfc8837#name-dscp-mappings.
Libwebrtc is only using AF41/AF42 for video high/very high.

achristensen07 commented on Source/WebKit/Platform/spi/Cocoa/NWSPI.h:
```diff
@@ -46,6 +46,14 @@ void nw_parameters_set_is_known_tracker(nw_parameters_t, bool is_known_tracker);
 void nw_parameters_allow_sharing_port_with_listener(nw_parameters_t, nw_listener_t);
 #endif // HAVE(NWPARAMETERS_TRACKER_API)
 
+#define SO_TC_BK_SYS 100
```
I believe these are defined in sys/socket.h so shouldn't need to be redefined in an SPI header.

youennf replied:
The public socket.h does not seem to have it.
It has service class/type but not traffic class, which is at a lower level.

achristensen07 reviewed (approved):
Approved once comments are responded to

youennf committed:
Set the traffic class of the NW WebRTC backend based on WebRTC backend DSCP information
https://bugs.webkit.org/show_bug.cgi?id=254072
rdar://106722223

Reviewed by Alex Christensen.

Set traffic class based on provided DSCP values for TCP and UDP nw sockets.

* Source/WebKit/NetworkProcess/webrtc/NetworkRTCTCPSocketCocoa.mm:
(WebKit::NetworkRTCTCPSocketCocoa::setOption):
* Source/WebKit/NetworkProcess/webrtc/NetworkRTCUDPSocketCocoa.mm:
(WebKit::NetworkRTCUDPSocketCocoaConnections::setOption):
* Source/WebKit/NetworkProcess/webrtc/NetworkRTCUtilitiesCocoa.h:
* Source/WebKit/NetworkProcess/webrtc/NetworkRTCUtilitiesCocoa.mm:
(WebKit::trafficClassFromDSCP):

Canonical link: https://commits.webkit.org/261917@main
